core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1647,6 +1647,14 @@ abstract class RDD[T: ClassTag](
     }
   }
 
+  /**
+   * :: Experimental ::
+   * Indicates that Spark must launch the tasks together for the current stage.
+   */
+  @Experimental
+  @Since("2.4.0")
+  def barrier(): RDDBarrier[T] = withScope(new RDDBarrier[T](this))
```

Comment:
scheduling from seems to have a very hard requirement that the number of partitions is less than the number of available task slots.  It seems really hard for users to get this right.  Eg., if I just do

`sc.textFile(...).barrier().mapPartitions()`

the number of partitions is based on the hdfs input splits.  I see lots of users getting confused by this -- it'll work sometimes, won't work other times, and they won't know why.  Should there be some automatic repartitioning based on cluster resources?  Or at least an api which lets users do this?  Even `repartition()` isn't great here, because users dont' want to think about cluster resources.

Reply:
Em, thanks for raising this question. IMO we indeed require users be aware of how many tasks they may launch for a barrier stage, and tasks may exchange internal data between each other in the middle, so users really care about the task numbers. I agree it shall be very useful to enable specify the number of tasks in a barrier stage, maybe we can have `RDDBarrier.coalesce(numPartitions: Int)` to enforce the number of tasks to be launched together in a barrier stage.

Reply:
I opened https://issues.apache.org/jira/browse/SPARK-24941 for this.